app/src/main/java/org/schabi/newpipe/player/PlayerService.kt

```diff
@@ -0,0 +1,207 @@
+            }
+            return field
+        }
+
+    val sessionConnector: MediaSessionConnector?
+        get() = mediaBrowserConnector?.sessionConnector
+
+    /*//////////////////////////////////////////////////////////////////////////
+    // Service's LifeCycle
+    ////////////////////////////////////////////////////////////////////////// */
+    override fun onCreate() {
+        super.onCreate()
+
+        if (DEBUG) {
+            Log.d(TAG, "onCreate() called")
+        }
+        Localization.assureCorrectAppLanguage(this)
+        ThemeHelper.setTheme(this)
+    }
+
+    override fun onStartCommand(
+        intent: Intent,
+        flags: Int,
+        startId: Int,
+    ): Int {
+        if (DEBUG) {
+            Log.d(
+                TAG,
+                "onStartCommand() called with: intent = [" + intent +
+                    "], flags = [" + flags + "], startId = [" + startId + "]",
+            )
+        }
+
+        if (Intent.ACTION_MEDIA_BUTTON == intent.action && (player.playQueue == null)) {
+            /*
+            No need to process media button's actions if the player is not working, otherwise
+            the player service would strangely start with nothing to play
+            Stop the service in this case, which will be removed from the foreground and its
+            notification cancelled in its destruction
+             */
+            stopSelf()
+            return START_NOT_STICKY
+        }
+
+        player.handleIntent(intent)
+
+        return START_NOT_STICKY
+    }
+
+    fun stopForImmediateReusing() {
+        if (DEBUG) {
+            Log.d(TAG, "stopForImmediateReusing() called")
+        }
+
+        if (!player.exoPlayerIsNull()) {
+            // Releases wifi & cpu, disables keepScreenOn, etc.
+            // We can't just pause the player here because it will make transition
+            // from one stream to a new stream not smooth
+            player.smoothStopForImmediateReusing()
+        }
+    }
+
+    override fun onTaskRemoved(rootIntent: Intent) {
+        super.onTaskRemoved(rootIntent)
+        if (!player.videoPlayerSelected()) {
+            return
+        }
+        onDestroy()
+        // Unload from memory completely
+        Runtime.getRuntime().halt(0)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        if (DEBUG) {
+            Log.d(TAG, "destroy() called")
+        }
+
+        cleanup()
+
+        mediaBrowserConnector?.release()
+        mediaBrowserConnector = null
+
+        compositeDisposableLoadChildren.clear()
+    }
+
+    private fun cleanup() {
+        player.destroy()
+    }
+
+    fun stopService() {
+        cleanup()
+        stopSelf()
+    }
+
+    override fun attachBaseContext(base: Context) {
+        super.attachBaseContext(AudioServiceLeakFix.preventLeakOf(base))
+    }
+
+    override fun onBind(intent: Intent): IBinder? {
+        if (SERVICE_INTERFACE == intent.action) {
```

Comment:
Add a comment here to explain what this means

Reply:
@haggaie Can you add details here.

Reply:
I think the goal here was to send `MediaBrowserServiceCompat` messages to the base class, while keeping the existing custom binder `PlayerService.LocalBinder` interface for the existing messages.